File: app/src/main/java/com/amazon/tv/leanbacklauncher/inputs/InputsTextView.java
package com.amazon.tv.leanbacklauncher.inputs;

import android.content.Context;
import android.content.res.Resources;
import android.text.Layout;
import android.util.AttributeSet;
import android.util.TypedValue;

import androidx.appcompat.widget.AppCompatTextView;

import com.amazon.tv.leanbacklauncher.R;

public final class InputsTextView extends AppCompatTextView {
    private TypedValue mPaddingDefault = new TypedValue();
    private TypedValue mPaddingWrapped = new TypedValue();
    private Resources mRes;
    private boolean mTextChanged;
    private TypedValue mTextSizeDefault = new TypedValue();
    private TypedValue mTextSizeWrapped = new TypedValue();

    public InputsTextView(Context context) {
        super(context);
        init(context);
    }

    public InputsTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public InputsTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        this.mRes = context.getResources();
        this.mRes.getValue(R.dimen.input_banner_text_bottom_padding, this.mPaddingDefault, false);
        this.mRes.getValue(R.dimen.input_banner_text_bottom_padding_wrapped, this.mPaddingWrapped, false);
        this.mRes.getValue(R.dimen.input_banner_label_text_size, this.mTextSizeDefault, false);
        this.mRes.getValue(R.dimen.input_banner_label_text_size_wrapped, this.mTextSizeWrapped, false);
    }

    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (this.mTextChanged) {
            this.mTextChanged = false;
            setMaxLines(1);
            setTextSize(this.mTextSizeDefault);
            setPaddingBottom(this.mPaddingDefault);
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
            if (isEllipsized()) {
                setTextSize(this.mTextSizeWrapped);
                super.onMeasure(widthMeasureSpec, heightMeasureSpec);
                if (isEllipsized()) {
                    setMaxLines(2);
                    setPaddingBottom(this.mPaddingWrapped);
                    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
                    return;
                }
                return;
            }
            return;
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    private void setTextSize(TypedValue value) {
        setTextSize((value.data >> 0) & 15, TypedValue.complexToFloat(value.data));
    }

    private void setPaddingBottom(TypedValue value) {
        setPadding(getPaddingLeft(), getPaddingTop(), getPaddingRight(), TypedValue.complexToDimensionPixelSize(value.data, this.mRes.getDisplayMetrics()));
    }

    public void setText(CharSequence text, BufferType type) {
        this.mTextChanged = true;
        requestLayout();
        super.setText(text, type);
    }

    private boolean isEllipsized() {
        Layout l = getLayout();
        if (l == null) {
            return false;
        }
        int lines = l.getLineCount();
        return lines > 0 && l.getEllipsisCount(lines - 1) > 0;
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/inputs/InputsTextView.kt
package com.amazon.tv.leanbacklauncher.inputs

import android.content.Context
import android.content.res.Resources
import android.util.AttributeSet
import android.util.TypedValue
import androidx.appcompat.widget.AppCompatTextView
import com.amazon.tv.leanbacklauncher.R

class InputsTextView : AppCompatTextView {
    private val mPaddingDefault = TypedValue()
    private val mPaddingWrapped = TypedValue()
    private var mRes: Resources? = null
    private var mTextChanged = false
    private val mTextSizeDefault = TypedValue()
    private val mTextSizeWrapped = TypedValue()

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context)
    }

    private fun init(context: Context) {
        mRes = context.resources
        mRes?.let {
            it.getValue(R.dimen.input_banner_text_bottom_padding, mPaddingDefault, false)
            it.getValue(R.dimen.input_banner_text_bottom_padding_wrapped, mPaddingWrapped, false)
            it.getValue(R.dimen.input_banner_label_text_size, mTextSizeDefault, false)
            it.getValue(R.dimen.input_banner_label_text_size_wrapped, mTextSizeWrapped, false)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (mTextChanged) {
            mTextChanged = false
            maxLines = 1
            setTextSize(mTextSizeDefault)
            setPaddingBottom(mPaddingDefault)
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
            if (isEllipsized) {
                setTextSize(mTextSizeWrapped)
                super.onMeasure(widthMeasureSpec, heightMeasureSpec)
                if (isEllipsized) {
                    maxLines = 2
                    setPaddingBottom(mPaddingWrapped)
                    super.onMeasure(widthMeasureSpec, heightMeasureSpec)
                    return
                }
                return
            }
            return
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    private fun setTextSize(value: TypedValue) {
        setTextSize(value.data shr 0 and 15, TypedValue.complexToFloat(value.data))
    }

    private fun setPaddingBottom(value: TypedValue) {
        setPadding(
            paddingLeft,
            paddingTop,
            paddingRight,
            TypedValue.complexToDimensionPixelSize(value.data, mRes!!.displayMetrics)
        )
    }

    override fun setText(text: CharSequence, type: BufferType) {
        mTextChanged = true
        requestLayout()
        super.setText(text, type)
    }

    private val isEllipsized: Boolean
        get() {
            val l = layout ?: return false
            val lines = l.lineCount
            return lines > 0 && l.getEllipsisCount(lines - 1) > 0
        }
}